### PR TITLE
Sb 10092

### PR DESCRIPTION
**Pull request by amit-tarento**

### CODE REVIEW

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 1]
```diff
@@ -70,20 +96,238 @@ public void onReceive(Request request) throws Throwable {
...
     }
     sender().tell(response, sender());
   }
 
+  private void validateTopics(Set<String> topics, String frameworkId) {
+    Set<String> frameworkTopics = getRelatedFrameworkTopics(frameworkId);
+    topics.forEach(
+        name -> {
+          if (frameworkTopics.add(name)) {
+            throwClientErrorException(
```
Comment (amolnin) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: collect all invalid topic names in list/set and then display error together

Comment (amolnin) [reply actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: This way user can know all the errors in one go.

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 2]
```diff
@@ -70,20 +96,238 @@ public void onReceive(Request request) throws Throwable {
...
+              s -> {
+                if (JsonKey.TOPIC.equalsIgnoreCase((String) s.get(JsonKey.CODE))) {
+                  terms.addAll((List<Map<String, Object>>) s.get(JsonKey.TERMS));
+                }
+              });
+      }
+    }
+    Set<String> topics = new HashSet<>();
+    for (Map<String, Object> term : terms) {
+      List<Map<String, Object>> childrens = (List<Map<String, Object>>) term.get(JsonKey.CHILDREN);
```
Comment (amolnin) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: childrens -> children (this is already plural)

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 3]
```diff
@@ -70,20 +96,238 @@ public void onReceive(Request request) throws Throwable {
...
+      if (CollectionUtils.isNotEmpty(childrens)) {
+        for (Map<String, Object> child : childrens) {
+          topics.add((String) child.get(JsonKey.NAME));
```
Comment (amolnin) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: topcis could be only till 2 levels - if it is recursive - this only takes care of 2 levels. Hence asking.

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 4]
```diff
@@ -70,20 +96,238 @@ public void onReceive(Request request) throws Throwable {
...
+  }
+
+  @SuppressWarnings("unchecked")
+  private void validateDialCodesWithReservedDialCodes(
+      Set<String> dialCodes, Map<String, Object> textBookdata) {
+    List<String> reservedDialCodes = (List<String>) textBookdata.get(JsonKey.RESERVED_DIAL_CODES);
+    if (CollectionUtils.isNotEmpty(reservedDialCodes)) {
+      dialCodes.forEach(
+          dialCode -> {
+            if (!reservedDialCodes.contains(dialCode)) {
```
Comment (amolnin) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: again, we shud collect all invalid dialcodes

**Comment by amolnin** [comment]
Unit-tests need to be extended or written - for it to be merged

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 5]
```diff
@@ -461,11 +712,38 @@ private void populateNodeModified(
...
+                  put(JsonKey.DIAL_CODES, dCodes);
+                }
+                if (CollectionUtils.isNotEmpty(topics)) {
+                  List<String> topicList = new ArrayList<>();
+                  topics.forEach(
+                      s -> {
+                        topicList.add(s.trim());
+                      });
+                  put(JsonKey.TOPIC, topicList);
+                }
```
Comment (amolnin) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: above 2 blocks are exactly identical - can we do away with duplication?

Comment (amolnin) [reply actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: Use following expression as splitter
\\s*,\\s*
u can get rid of extra spaces - in split function itself.

Comment (amit-tarento) [reply actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: this is not working in few scenario.

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 6]
```diff
@@ -70,20 +96,248 @@ public void onReceive(Request request) throws Throwable {
...
+    InputStream inputStream = new ByteArrayInputStream(byteArray);
+    Map<String, Object> resultMap = readAndValidateCSV(inputStream);
+    Set<String> dialCodes = (Set<String>) resultMap.get(JsonKey.DIAL_CODES);
+    resultMap.remove(JsonKey.DIAL_CODES);
+
+    Set<String> topics = (Set<String>) resultMap.get(JsonKey.TOPICS);
+    resultMap.remove(JsonKey.TOPICS);
+
+    String tbId = (String) request.get(TEXTBOOK_ID);
+    Map<String, Object> textBookdata = getTextbook(tbId);
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: textbookMap (or) textbookData

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 7]
```diff
@@ -70,20 +96,248 @@ public void onReceive(Request request) throws Throwable {
...
+        if (null != csvFileParser) csvFileParser.close();
+      } catch (IOException e) {
+        ProjectLogger.log(
+            "TextbookTocActor:readAndValidateCSV : Exception occurred while closing stream");
+      }
+    }
+    return result;
+  }
+
+  private void validateOrCodeRequiredAndQrCode(Map<String, Object> recordMap) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: validateQrCodeRequiredAndQrCode

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 8]
```diff
@@ -70,20 +96,255 @@ public void onReceive(Request request) throws Throwable {
...
     }
     sender().tell(response, sender());
   }
 
+  private void validateTopics(Set<String> topics, String frameworkId) {
+    Set<String> frameworkTopics = getRelatedFrameworkTopics(frameworkId);
+    Set<String> invalidTopics = new HashSet<>();
+    topics.forEach(
+        name -> {
+          if (frameworkTopics.add(name)) {
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: Use an alternate Set method to check if a topic exists or not.

File: actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java [hunk 9]
```diff
@@ -70,20 +96,255 @@ public void onReceive(Request request) throws Throwable {
...
+      if (CollectionUtils.isNotEmpty(invalidDialCodes)) {
+        throwClientErrorException(
+            ResponseCode.errorDialCodeNotReservedForTextBook,
+            MessageFormat.format(
+                ResponseCode.errorDialCodeNotReservedForTextBook.getErrorMessage(),
+                StringUtils.join(invalidDialCodes, ','),
+                textBookdata.get(JsonKey.IDENTIFIER)));
+      }
+    } else {
+      throwClientErrorException(
```
Comment (bvinayakumar) [inline actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java]: This error to be thrown only if dialCodes is not empty.

**Pull request merged**